File: app/services/email_service.py
"""Email sending service — async SMTP with PDF attachment.

Uses aiosmtplib for non-blocking SMTP operations.
"""
import asyncio
import ssl
from email.mime.multipart import MIMEMultipart
from email.mime.base import MIMEBase
from email.mime.text import MIMEText
from email import encoders



class EmailConfig:
    def __init__(self, host: str, port: int, username: str, password: str,
                 use_ssl: bool = True, **kwargs):
        self.host = host
        self.port = port
        self.username = username
        self.password = password
        self.use_ssl = use_ssl
        # Store from_name for backward compatibility (used by API layer)
        self.from_name = kwargs.get("from_name", "")

    def is_configured(self) -> bool:
        return bool(self.host and self.username and self.password)


async def send_statement_email(
    config: EmailConfig,
    to_email: str,
    bank_name: str,
    account_id: str,
    pdf_bytes: bytes,
) -> dict:
    """Send an email with the PDF attachment using SMTP (async).

    The email appears to come from the bank (noreply@bankname.com).
    Uses synchronous smtplib wrapped in asyncio.to_thread() to avoid blocking
    the event loop. For truly async SMTP, aiosmtplib is also an option.
    """
    if not config.is_configured():
        return {"status": "error", "message": "Email is not configured."}

    def _mask_account(acct: str) -> str:
        if not acct or len(acct) < 4:
            return acct or ""
        return "X" * (len(acct) - 4) + acct[-4:]

    masked_acct = _mask_account(account_id)

    msg = MIMEMultipart("mixed")
    # Use bank name as the sender display name
    msg["From"] = f"{bank_name.upper()} Statement"
    msg["To"] = to_email
    msg["Subject"] = f"Account Statement from {bank_name} - {masked_acct}"

    # HTML body with bank branding
    html_body = f"""<!DOCTYPE html>
<html><head><meta charset='UTF-8'><meta name='viewport' content='width=device-width, initial-scale=1.0'></head>
<body style='margin:0;padding:0;background:#f4f6f8;font-family:Arial,Helvetica,sans-serif;'>
<table width='100%' cellpadding='0' cellspacing='0' style='background:#f4f6f8;padding:20px;'>
<tr><td align='center'>
<table width='600' cellpadding='0' cellspacing='0' style='background:#fff;border-radius:4px;overflow:hidden;box-shadow:0 2px 8px rgba(0,0,0,0.1);'>

<!-- HDFC Bank Header Banner -->
<tr><td style='background:#004B8D;padding:20px 30px;text-align:left;'>
<table width='100%' cellpadding='0' cellspacing='0'><tr>
<td style='width:50%;'>
<div style='font-size:24px;font-weight:bold;color:#fff;letter-spacing:1px;'>HDFC BANK</div>
<div style='font-size:11px;color:#b3d4ff;margin-top:4px;'>We understand your world</div>
</td>
<td style='width:50%;text-align:right;'>
<div style='font-size:10px;color:#b3d4ff;'>Secure Document Delivery</div>
</td>
</tr></table>
</td></tr>

<!-- Body -->
<tr><td style='padding:30px;'>
<p style='font-size:13px;color:#333;line-height:1.6;margin:0 0 12px 0;'>As per request, please find attached your <strong>{bank_name}</strong> account statement for the account ending with <strong>{masked_acct[-4:]}</strong>.</p>

<!-- Account Details Table -->
<table width='100%' cellpadding='8' cellspacing='0' style='font-size:12px;color:#333;border:1px solid #ddd;margin:16px 0;'>
<tr style='background:#f0f5fa;'><td style='width:150px;font-weight:600;border-bottom:1px solid #ddd;'>Account Number</td><td style='border-bottom:1px solid #ddd;'>{masked_acct}</td></tr>
<tr><td style='font-weight:600;border-bottom:1px solid #ddd;'>Statement Period</td><td style='border-bottom:1px solid #ddd;'>Latest Statement</td></tr>
<tr style='background:#f0f5fa;'><td style='font-weight:600;border-bottom:1px solid #ddd;'>Document Type</td><td style='border-bottom:1px solid #ddd;'>e-Statement (PDF)</td></tr>
</table>

<p style='font-size:12px;color:#555;line-height:1.5;margin:16px 0;'>The attached PDF is password protected with the last 4 digits of your account number.</p>

<p style='font-size:12px;color:#555;line-height:1.5;margin:16px 0;'>If you have any queries, please contact our 24x7 PhoneBanking at <strong>1800 266 4332</strong> or visit your nearest {bank_name} branch.</p>

<p style='font-size:12px;color:#555;line-height:1.5;margin:16px 0 0 0;'>Warm Regards,<br/><strong>Customer Service Team</strong><br/>{bank_name}</p>
</td></tr>

<!-- Footer -->
<tr><td style='background:#004B8D;padding:16px 30px;text-align:center;border-top:2px solid #003366;'>
<p style='font-size:10px;color:#b3d4ff;margin:0 0 4px 0;'>This is a system generated e-mail. Please do not reply to this message.</p>
<p style='font-size:10px;color:#b3d4ff;margin:0 0 4px 0;'>HDFC Bank Ltd. | {bank_name} House, 165-166, Backbay Reclamation, Mumbai 400 020</p>
<p style='font-size:10px;color:#b3d4ff;margin:0;'>&copy; 2024 {bank_name} Ltd. All rights reserved.</p>
</td></tr>

</table>
</td></tr></table></body></html>"""

    msg.attach(MIMEText(html_body, "html"))

    # Attach PDF
    attachment = MIMEBase("application", "octet-stream")
    attachment.set_payload(pdf_bytes)
    encoders.encode_base64(attachment)
    filename = f"statement-{account_id}.pdf"
    attachment.add_header("Content-Disposition", f"attachment; filename={filename}")
    msg.attach(attachment)

    def _send_sync():
        import smtplib
        if config.port == 465:
            # Port 465 uses implicit SSL/TLS
            context = ssl.create_default_context()
            with smtplib.SMTP_SSL(config.host, config.port, context=context) as server:
                server.login(config.username, config.password)
                server.send_message(msg)
        else:
            # Port 587 and others use STARTTLS (connect plain, then upgrade)
            with smtplib.SMTP(config.host, config.port) as server:
                server.ehlo()
                server.starttls(context=ssl.create_default_context())
                server.ehlo()
                server.login(config.username, config.password)
                server.send_message(msg)

    try:
        # Run synchronous SMTP in a thread to avoid blocking event loop
        await asyncio.to_thread(_send_sync)
        return {
            "status": "ok",
            "message": f"Email sent successfully to {to_email}",
            "to": to_email,
            "subject": msg["Subject"],
            "attachment": filename,
        }
    except Exception as e:
        return {"status": "error", "message": f"Failed to send email: {str(e)}"}

File: backend/src/main/java/com/bankeditor/service/EmailService.java
package com.bankeditor.service;

import jakarta.mail.internet.InternetAddress;
import jakarta.mail.internet.MimeMessage;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.mail.javamail.JavaMailSenderImpl;
import org.springframework.mail.javamail.MimeMessageHelper;
import org.springframework.stereotype.Service;

import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Properties;

@Service
public class EmailService {

    private volatile JavaMailSenderImpl mailSender;
    private volatile EmailConfig currentConfig;

    public static class EmailConfig {
        public String host;
        public int port;
        public String username;
        public String password;
        public boolean useSsl;
        public String fromName;

        public boolean isConfigured() {
            return host != null && !host.isEmpty()
                && username != null && !username.isEmpty()
                && password != null && !password.isEmpty();
        }
    }

    public boolean isConfigured() {
        return currentConfig != null && currentConfig.isConfigured();
    }

    public Map<String, Object> getConfigStatus() {
        Map<String, Object> status = new LinkedHashMap<>();
        status.put("configured", isConfigured());
        if (currentConfig != null) {
            status.put("host", currentConfig.host != null ? currentConfig.host : "");
            status.put("port", currentConfig.port);
            status.put("username", currentConfig.username != null ? currentConfig.username : "");
            status.put("fromName", currentConfig.fromName != null ? currentConfig.fromName : "Bank Statement Service");
        }
        return status;
    }

    public void configure(EmailConfig config) {
        this.currentConfig = config;
        JavaMailSenderImpl impl = new JavaMailSenderImpl();
        impl.setHost(config.host);
        impl.setPort(config.port);
        impl.setUsername(config.username);
        impl.setPassword(config.password);

        Properties props = new Properties();
        props.put("mail.smtp.auth", "true");
        if (config.port == 465) {
            // Port 465 uses direct SSL
            props.put("mail.smtp.ssl.enable", "true");
        } else {
            // Port 587 uses STARTTLS (upgrade from plain to encrypted)
            props.put("mail.smtp.starttls.enable", "true");
            props.put("mail.smtp.starttls.required", "true");
        }
        props.put("mail.smtp.connectiontimeout", "10000");
        props.put("mail.smtp.timeout", "10000");
        props.put("mail.debug", "false");
        impl.setJavaMailProperties(props);

        this.mailSender = impl;
    }

    public Map<String, Object> sendStatementEmail(String toEmail, String bankName, String accountHolder,
                                                   String accountId, byte[] pdfBytes, String password) {
        Map<String, Object> result = new LinkedHashMap<>();

        if (!isConfigured()) {
            result.put("status", "error");
            result.put("message", "Email is not configured. Please set SMTP settings first.");
            return result;
        }
        if (mailSender == null) {
            result.put("status", "error");
            result.put("message", "Mail sender not initialized. Configure email first.");
            return result;
        }

        try {
            MimeMessage message = mailSender.createMimeMessage();
            MimeMessageHelper helper = new MimeMessageHelper(message, true);

            String maskedAccount = maskAccount(accountId);
            String fromAddress = currentConfig.username;
            String fromName = (currentConfig.fromName != null && !currentConfig.fromName.isEmpty())
                ? currentConfig.fromName : "Bank Statement Service";

            try {
                helper.setFrom(new InternetAddress(fromAddress, fromName, "UTF-8"));
            } catch (Exception e) {
                helper.setFrom(fromAddress);
            }
            helper.setTo(toEmail);
            helper.setSubject("Your " + bankName + " Account Statement - " + maskedAccount);

            String htmlBody = "<!DOCTYPE html>" +
                "<html><body style='font-family:Arial,sans-serif;padding:20px;'>" +
                "<div style='max-width:600px;margin:0 auto;border:1px solid #e0e0e0;border-radius:8px;overflow:hidden;'>" +
                "<div style='background:linear-gradient(135deg,#0a1628,#1a1a3e);padding:24px;text-align:center;'>" +
                "<h2 style='color:#fff;margin:0;'>" + bankName + "</h2>" +
                "<p style='color:#8899bb;font-size:13px;margin:4px 0 0 0;'>Secure Statement Delivery</p>" +
                "</div>" +
                "<div style='padding:24px;'>" +
                "<p style='font-size:15px;color:#333;'>Dear <strong>" + accountHolder + "</strong>,</p>" +
                "<p style='font-size:14px;color:#555;line-height:1.6;'>Please find attached your account statement for <strong>" + bankName + "</strong> (Account: " + maskedAccount + ").</p>" +
                "<div style='background:#f0f7ff;border:1px solid #cce5ff;border-radius:8px;padding:16px;margin:20px 0;'>" +
                "<p style='font-size:13px;color:#555;margin:0 0 8px 0;'><strong>Statement Details:</strong></p>" +
                "<table width='100%' cellpadding='4' cellspacing='0' style='font-size:13px;color:#555;'>" +
                "<tr><td style='color:#999;width:120px;'>Account Holder:</td><td style='font-weight:600;'>" + accountHolder + "</td></tr>" +
                "<tr><td style='color:#999;'>Account Number:</td><td style='font-weight:600;'>" + maskedAccount + "</td></tr>" +
                "<tr><td style='color:#999;'>PDF Password:</td><td style='font-weight:600;color:#2563eb;'>" + password + "</td></tr>" +
                "</table></div>" +
                "<p style='font-size:13px;color:#666;line-height:1.5;'>The statement PDF is password-protected. Use the last 4 digits of your account number (<strong>" + password + "</strong>) to open it.</p>" +
                "<p style='font-size:13px;color:#999;line-height:1.5;margin:20px 0 0 0;'>This is an automated message from " + bankName + ". Please do not reply to this email.</p>" +
                "</div>" +
                "<div style='background:#f8f9fb;padding:16px;text-align:center;border-top:1px solid #e0e0e0;'>" +
                "<p style='font-size:12px;color:#999;margin:0;'>&copy; 2024 " + bankName + ". All rights reserved.</p>" +
                "</div>" +
                "</div></body></html>";

            helper.setText(htmlBody, true);

            String filename = "statement-" + accountId + ".pdf";
            helper.addAttachment(filename, new ByteArrayResource(pdfBytes));

            mailSender.send(message);
            result.put("status", "ok");
            result.put("message", "Email sent successfully to " + toEmail);
            result.put("to", toEmail);
            result.put("subject", message.getSubject());
            result.put("attachment", filename);
        } catch (Exception e) {
            result.put("status", "error");
            result.put("message", "Failed to send email: " + e.getMessage());
        }
        return result;
    }

    private String maskAccount(String accountId) {
        if (accountId == null || accountId.length() < 4) return accountId;
        String lastFour = accountId.substring(accountId.length() - 4);
        String masked = "X".repeat(Math.max(0, accountId.length() - 4));
        return masked + lastFour;
    }
}
